### DISCUSSION TITLE
Physics-informed PySR

### DISCUSSION THREAD

**Discussion by AKHCE** [MilesCranmer/PySR#568] [Ideas]
Hi,
I am wondering if anyone has tried to find an analytical approximate solution for a PDE using PySR. I mean, it seems feasible to apply a similar methodology to PINNs for PySR; specifically, in each iteration in PySR, the error of each equation proposed by PySR is computed through the sum of errors associated with the PDE, BC, and IC. I think this approach can be called Physics-Informed PySR. I am curious, have you known/come across any study about Physics-Informed PySR?
Any idea/viewpoint is much appreciated,
Thanks

**Comment by pukpr** [comment]
I'm sure it can be done. In Eureqa, it was possible to express the following

$\frac{{d^2 f(t)}}{{d t^2 }} + \omega^2 f(t) = F(t)$

The built-in function D is a derivative (2 means 2nd derivative), and sma is a smoothing averager. soi is a time-series data

![image](https://github.com/MilesCranmer/PySR/assets/2855758/691aeef0-3c99-479c-92fe-22be51d38032)

**Reply by AKHCE** [reply]
Thank you so much. Basically, what I infer from the interface is that Eureqa is capable of searching for analytical solution satisfying an ODE or algebraic equation, pretty much interesting! 
I am wondering if you know how Eureqa accepts initial condition/boundary condition?
Thanks a ton for sharing

**Reply by pukpr** [reply]
Discovering initial conditions is implicit in the process is the best way to put it.  As always, the big challenge is not in IC/BC but in discovering what a possible unknown forcing function $F(t)$ is.  Nature does not usually work by providing perfect impulse or delta functions whereby one can recover the natural response to a differential equation. It's typically messier than that, with a forcing that may be wandering around, so the response is not necessarily stationary.  IMO, that's the big hurdle in Physics-Informed PySR, in that you will only extract the physics if you know the inputs and outputs well.  Consider if you had an audio sample that you think had been filtered (often a high-order diffEq) -- you could never guess the physics unless you had the raw input.

The practical advice I would give is be careful about providing a derivative along with an integrator.  The integrator can undo the results of the differentiation so that it will recover $f(t)$ and it will converge to a perfect match.

**Comment by MilesCranmer** [comment]
Thanks @pukpr and sorry for late reply @AKHCE!

You should check out https://github.com/MilesCranmer/PySR/discussions/401 for how to do this, which basically finds $f(x_1, x_2, ...)$ such that
$$\frac{\partial f(x_1, x_2, ...)}{\partial x_1} = y.$$

Note that there is no built-in second-order autodiff, only first-order. (My hope is that Enzyme.jl will eventually give us this for free). A temporary workaround would be to use finite difference approximation for second-order.

Also check out: https://github.com/MilesCranmer/PySR/discussions/299 https://github.com/MilesCranmer/PySR/discussions/414 https://github.com/MilesCranmer/PySR/discussions/447 https://github.com/MilesCranmer/PySR/discussions/548.

**Reply by AKHCE** [reply]
Thanks a lot

**Comment by pukpr** [comment]
I have a set of what I would categorize as unsolved geophysics time series problems that I am modeling. The solutions I have found are all physics-informed but not necessarily well-accepted, even after publishing.  Over the years, I have experimented with using symbolic regression to see if it could discover the same solutions, and I did find that Eureqa was very promising in being able to get close.  I don't know what it is but Eureqa had some sort of special sauce to it; unfortunately it is out of commission these days.

But, if anyone is interested the problems that I am working on that could benefit from symbolic regression experimentation are:
Chandler wobble -- https://en.wikipedia.org/wiki/Chandler_wobble
Quasi-Biennial Oscillation (QBO) -- https://en.wikipedia.org/wiki/Quasi-biennial_oscillation
El Nino Southern Oscillation (ENSO) -- https://en.wikipedia.org/wiki/El_Ni%C3%B1o%E2%80%93Southern_Oscillation
Atlantic Multidecadal Oscillation (AMO) -- https://en.wikipedia.org/wiki/Atlantic_multidecadal_oscillation
Indian Ocean Dipole (IOD) -- https://en.wikipedia.org/wiki/Indian_Ocean_Dipole
Pacific Decadal Oscillation (PDO) -- https://en.wikipedia.org/wiki/Pacific_decadal_oscillation
North Atlantic Oscillation (NAO) -- https://en.wikipedia.org/wiki/North_Atlantic_oscillation

There is a commonality to the physics-modeling approach I am using on all of these and so there is a strong possibility of cross-validation.

However, the sticky point in getting something like PySR to match is in first coercing it to it lock-in to the formulation I am finding. That''s why I haven't been using PySR as much as using custom software.

My suggestion as to how to proceed is to instead of providing the real data of QBO (for example) for PySR to  fit to, instead to try to find out if PySR could fit to a **model** of QBO, which would be free of noise or other anomalies.

This would be huge if even a few of these problems could be solved, as the reality is that geophysicists and climate scientists really have little faith in their own models for the phenomena.   Google, NVIDIA, and Huawei are spending millions $$ in research on the climate aspects to these time series, clearly concentrating on using neural networks to find something.   However, unless they get the physics and especially the external forcings right, I don't think the NNs are going to find the solutions.  That's because the solutions aren't found solely within the data but as a consequence of manipulating the external factors.  And that's why I think symbolic regression approaches will work better.

---
Example of QBO PySR fit, a tropical/synodic tidal factor will emerge in addition to the draconic/nodal factor supplied
https://github.com/orgs/azimuth-project/discussions/10#discussion-6397169

**Reply by pukpr** [reply]
Example of AMO model PySR fit
![image](https://github.com/user-attachments/assets/102851b7-f216-4fff-8e20-57fc97e090db)

`
Complexity  Loss       Score     Equation   
1           2.145e+00  1.594e+01  y = -0.1587         
3           2.093e+00  1.228e-02  y = sin(sin(x₀))    
4           1.426e+00  3.832e-01  y = sin(x₁ / -0.4348)   
5           1.376e+00  3.627e-02  y = cos(-1.9521 / exp(x₀))         
6           1.202e+00  1.349e-01  y = sin((x₀ / 0.403) - 0.90448)    
8           1.031e+00  7.661e-02  y = (sin((x₀ / 0.43131) - 0.96015) * 1.662)         
9           7.839e-01  2.743e-01  y = (sin(x₁ / -0.43495) - cos(2.3502 * x₀))     
10          6.512e-01  1.854e-01  y = (cos(1.9457 / exp(x₀)) + sin(x₁ / -0.4348))  
11          4.573e-01  3.534e-01  y = (sin(x₁ / -0.43492) - cos((x₀ - -0.26261) * -2.4435))     
13          2.772e-01  2.503e-01  y = ((sin((x₀ / 0.43131) - 0.98787) * 1.662) + sin(x₁ / -0.43492))          
15          8.096e-02  6.155e-01  y = ((sin((x₁ + -0.23221) / -0.43531) - cos(0.65168 + (2.3256 * x₀))) * 1.4954)   
16          6.329e-02  2.462e-01  y = ((sin(sin((x₁ / -0.43531) - -0.52999)) - cos(0.65168 + (2.3502 * x₀))) * 1.5773)    
17          5.030e-02  2.296e-01  y = (((sin((x₁ / -0.43534) - -0.56788) - cos(0.73106 + (x₀ * 2.6611))) - 0.2069) * 1.4568)     
18          3.947e-02  2.426e-01  y = (((sin(sin((x₁ / -0.43532) - -0.56155)) - cos(0.7146 + (x₀ * 2.6198))) - 0.18281) * 1.5628)    
19          3.671e-02  7.237e-02  y = ((((sin((x₁ / -0.43534) - -0.56141) / 1.1484) - cos(0.71498 + (x₀ * 2.6208))) - 0.18246) * 1.5645)   

Sorry, but IMO, `12.021*cos(1.2555*x1+0.8716)` should have the same complexity as `cos(x1)`.  That's why symbolic regression struggles to find anything interesting.

Longer blog: post ; https://geoenergymath.com/2024/11/10/lunar-torque-controls-all/
